src/main/java/pl/joegreen/sergeants/simulator/GameMap.java

```diff
@@ -0,0 +1,240 @@
+                .forEach(tile -> tile.transfer(playerKilled.getOffender()));
+        return playerKilled;
+    }
+
+    public GameUpdateApiResponse getUpdate(int playerIndex) {
+        int[] mapDiff = getMapDiff(playerIndex);
+        int[] citiesDiff = getCitiesDiff();
+        int[] generals = getGenerals();
+        ScoreApiResponse[] scores = new ScoreApiResponse[0];
+        double[] stars = new double[0];
+        return new GameUpdateApiResponse(halfTurnCounter, mapDiff, citiesDiff, generals, 0, scores, stars);
+    }
+
+    private int[] getMapDiff(int playerIndex) {
+        List<Integer> mapDiff = new ArrayList<>();
+        //invalidate whole map diff and send a complete update because no need to save bandwidth
+        mapDiff.add(0);//this means client has zero correct values
+        mapDiff.add(tiles.length * 2 + 2); //this is number of corrected values
+        mapDiff.add(width);
+        mapDiff.add(height);
+
+        //army size
+        for (Tile tile : tiles) {
+            int armySize = isVisible(playerIndex, tile) ? tile.getArmySize() : 0;
+            mapDiff.add(armySize);
+        }
+
+        //terrain
+        for (Tile tile : tiles) {
+            boolean visible = isVisible(playerIndex, tile);
+            int terrain = tile.getTerrain(visible);
+            mapDiff.add(terrain);
+        }
+        return mapDiff.stream().mapToInt(Integer::intValue).toArray();
+    }
+
+    /**
+     * Returns all cities, visible and fogged. This is not a problem since bot will still receive tile as fogged obstacle (-4).
+     *
+     * @return list of tileIndex where a city is located
+     */
+    private int[] getCitiesDiff() {
+        List<Integer> allCities = Arrays.stream(tiles)
+                .filter(tile -> tile.getClass() == City.class)
+                .map(Tile::getTileIndex)
+                .collect(Collectors.toList());
+
+        allCities.add(0, allCities.size()); //second element = wrong
+        allCities.add(0, 0);//first element = zero correct
+        return allCities.stream().mapToInt(i -> i).toArray();
+    }
+
+    /**
+     * Returns all generals, visible and fogged. This is not a problem since bot will still receive tile as fogged (-3).
+     *
+     * @return list of tileIndex where general is, array index is player index
+     */
+    private int[] getGenerals() {
+        return Arrays.stream(tiles)
+                .filter(tile1 -> tile1.getClass().equals(General.class))
+                .sorted((t1, t2) -> t1.getPlayerIndex() - t2.getPlayerIndex())
+                .mapToInt(Tile::getTileIndex)
+                .toArray();
+    }
+
+    private boolean isVisible(int playerIndex, Tile tile) {
+        boolean selfTile = tile.getPlayerIndex() == playerIndex;
+        return selfTile || getSurroundingTiles(tile).stream().anyMatch(t -> t.getPlayerIndex() == playerIndex);
+    }
+
+    private List<Tile> getSurroundingTiles(Tile tile) {
```

Comment:
I have an idea and I think I could try to implement that change after we merge. I believe that GameMap could implement a GameState interface. It would probably make things generally simpler and if it kept tiles in a `Map<Position, Tile>` it could also use already implemented `GameState.getNeighbourPositions()` to perform what's implemented here. It would probably also simplify sending updates to players as there would be no need to create the low-level map/cities array diffs, we could just pass GameMap instance if it implemented `GameState`. 

There are some issues though as GameState contains some player-dependent things and that kind of breaks the idea:
*  int getPlayerIndex() 
* conversion from Tile to Field is also player dependent (as visibility of fields is different between players) 

That's where my idea fails for now, but maybe I will be able to find some workaround :-)